lib/services/access_list.go

```diff
@@ -218,47 +233,165 @@ func UnmarshalAccessListMember(data []byte, opts ...MarshalOption) (*accesslist.
+	MembershipOrOwnershipTypeInherited
+)
+
+// GetAccessListNestedLists returns the names of all nested lists that are direct Members of the provided Access List.
+func (a AccessListMembershipChecker) GetAccessListNestedLists(ctx context.Context, accessListName string) (nestedLists []string, err error) {
+	var pageToken string
+
+	for {
+		var members []*accesslist.AccessListMember
+
+		members, pageToken, err = a.members.ListAccessListMembers(ctx, accessListName, 0 /* default page size */, pageToken)
+		if err != nil {
+			return nil, trace.Wrap(err)
+		}
+
+		for _, member := range members {
+			if member.Spec.MembershipKind == accesslist.MembershipKindList {
+				nestedLists = append(nestedLists, member.GetName())
+			}
+		}
+
+		if pageToken == "" {
+			break
+		}
+	}
+
+	return nestedLists, nil
+}
+
+// IsAccessListOwner will determine both whether the provided tlsca.Identity is an owner for the accessList
+// and meets its ownership requirements.
+//
+// If ownership is inherited (through the tlsca.Identity having Membership in an Access List which is an Owner
+// of the provided accessList), the tlsca.Identity must also meet the membership requirements of the nested list.
+func (a AccessListMembershipChecker) IsAccessListOwner(ctx context.Context, identity tlsca.Identity, accessList *accesslist.AccessList) (MembershipOrOwnershipType, error) {
+	var ownershipType MembershipOrOwnershipType
+
+	// Is the supplied identity explicitly in the Owners list?
+	if slices.ContainsFunc(accessList.GetOwners(), func(owner accesslist.Owner) bool {
+		return owner.Name == identity.Username
+	}) {
+		ownershipType = MembershipOrOwnershipTypeExplicit
+	} else {
+		err := a.recursiveIsAccessListOwnerCheck(ctx, identity, accessList)
+		if err != nil {
+			if trace.IsNotFound(err) {
+				return MembershipOrOwnershipTypeNone, trace.AccessDenied("User '%s' is not an owner of list '%s'.", identity.Username, accessList.GetName())
+			}
+			return MembershipOrOwnershipTypeNone, trace.Wrap(err)
+		}
+		ownershipType = MembershipOrOwnershipTypeInherited
+	}
+
+	// Does the supplied Identity meet the ownership requirements?
+	if !UserMeetsRequirements(identity, accessList.Spec.OwnershipRequires) {
+		return MembershipOrOwnershipTypeNone, trace.AccessDenied("User '%s' is an owner, but does not have the roles or traits required to be an owner of list '%s'.", identity.Username, accessList.GetName())
+	}
+
+	// We've gotten through all the checks, so the user is a valid owner.
+	return ownershipType, nil
+}
+
+// recursiveIsAccessListOwnerCheck checks if supplied tlsca.Identity is an inherited owner of the accessList.
+// Ownership is inherited when the tlsca.Identity is a Member of an Access List, up to the maximum allowed depth,
+// which is an explicit Owner of the provided accessList.
+func (a AccessListMembershipChecker) recursiveIsAccessListOwnerCheck(ctx context.Context, identity tlsca.Identity, accessList *accesslist.AccessList) error {
```

Comment:
This is now duplication of the same logic in Hierarchy - can we get rid of this?

Reply:
Same as my reply above... in pursuit of keeping this PR from getting any bigger I'd much rather do a pass refactoring/removing these in a later PR, since the places they're used don't have access to `GetAccessLists`/`ListAccessLists`, and these instead make use of `GetAccessList`/`ListAccessListMembers`, so it'll take some refactoring.